### PR TITLE
Support nested fields in Elasticsearch Connector

### PR DESCRIPTION
**Pull request by zhenxiao**
Fix nested fields in Elasticsearch Connector
CC @nezihyigitbasi @wenleix @jwfcps @ebyhr @knikel

### CODE REVIEW

**Comment by zhenxiao** [comment]
Fix https://github.com/prestodb/presto/issues/12642
https://github.com/prestodb/presto/issues/12890

File: presto-elasticsearch/src/test/java/com/facebook/presto/elasticsearch/TestElasticsearchIntegrationSmokeTest.java [hunk 1]
```diff
@@ -90,4 +94,29 @@ public void testDescribeTable()
...
+        embeddedElasticsearchNode.getClient().admin().indices().refresh(refreshRequest(indexName)).actionGet();
+
+        MaterializedResult actualColumns = computeActual("SELECT name, fields.fielda, fields.fieldb FROM nested.data").toTestTypes();
+        MaterializedResult.Builder builder = resultBuilder(getQueryRunner().getDefaultSession(), VARCHAR, BIGINT, VARCHAR);
+        for (MaterializedRow row : actualColumns.getMaterializedRows()) {
+            builder.row(row.getField(0), row.getField(1), row.getField(2));
+        }
+
+        MaterializedResult actualResult = builder.build();
+        builder = resultBuilder(getQueryRunner().getDefaultSession(), VARCHAR, BIGINT, VARCHAR);
```
Comment (ebyhr) [inline presto-elasticsearch/src/test/java/com/facebook/presto/elasticsearch/TestElasticsearchIntegrationSmokeTest.java]: `BIGINT` looks wrong since the column is defined as integer in nested.data.json.
https://github.com/prestodb/presto/pull/12948/files#diff-4e0329725eb70ef7e5720bda1caca46cR20

How about simplifying these code?
```java
        MaterializedResult actualResult = computeActual("SELECT name, fields.fielda, fields.fieldb FROM nested.data").toTestTypes();
        MaterializedResult expectedResult = resultBuilder(getQueryRunner().getDefaultSession(), VARCHAR, INTEGER, VARCHAR)
                .row("nestfield", 32, "valueb")
                .build();
        assertEquals(actualResult, expectedResult);
```

**Comment by ebyhr** [comment]
In my understanding, this implementation allows only one level nest. Is it right?

File: presto-elasticsearch/src/test/resources/queryrunner/nested.data.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
+{
+    "tableName": "data",
+    "schemaName": "nested",
+    "host": "localhost",
+    "port": "9300",
+    "clusterName": "test",
+    "index": "data",
+    "type": "doc",
+    "columns": [
+
```
Comment (ebyhr) [inline presto-elasticsearch/src/test/resources/queryrunner/nested.data.json]: nit: Remove empty line

**Comment by zhenxiao** [comment]
thank you, @ebyhr 
I get comments addressed, and now the implementation is general for any level of nesting

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 3]
```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
...
+        Map.Entry<String, Object> first = fieldsMap.firstEntry();
+        String field = first.getKey();
+        if (field.contains(".")) {
+            String prefix = field.substring(0, field.lastIndexOf('.'));
+            Map<String, Object> value = new HashMap<>();
+            int size = field.split("\\.").length;
+            Iterator<String> iterator = fieldsMap.navigableKeySet().iterator();
+            while (iterator.hasNext()) {
+                String name = iterator.next();
+                if (name.split("\\.").length == size && name.startsWith(prefix)) {
```
Comment (wenleix) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: nit: What about `name.startsWith(prefix + ".")` ?

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 4]
```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
...
+                    value.put(name.substring(name.lastIndexOf('.') + 1), fieldsMap.get(name));
```
Comment (wenleix) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: Curious: Does this mechanism supports multiple level of JSON path ? (e.g. `a.b.c`)

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 5]
```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
...
+                    iterator.remove();
+                    continue;
+                }
+                break;
+            }
+            fieldsMap.put(prefix, value);
+            processFields(fieldsMap);
```
Comment (wenleix) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: Instead of using recursive, would it be easier to read and maintain if we can use loop? (e.g. 
```
while (!fieldsMap.isEmpty()) {
...
}
```

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 6]
```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
...
+        }
+        else {
+            for (Map.Entry<String, Object> entry : fieldsMap.entrySet()) {
```
Comment (wenleix) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: So does this mean if the first json path doesn't contain ".", all json paths don't contain "." ?

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 7]
```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
     private void extractFromSource(SearchHit hit)
     {
         Map<String, Object> map = hit.getSourceAsMap();
+        TreeMap<String, Object> fieldsMap = new TreeMap<>(new Comparator<String>()
```
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: nit
```java
TreeMap<String, Object> fieldsMap = new TreeMap<>((left, right) -> {
    int leftLength = left.split("\\.").length;
    int rightLength = right.split("\\.").length;
    if (leftLength == rightLength) {
        return left.compareTo(right);
    }
    return rightLength - leftLength;
});

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 8]
```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
...
+            {
+                int length1 = key1.split("\\.").length;
+                int length2 = key2.split("\\.").length;
+                if (length1 == length2) {
+                    return key1.compareTo(key2);
+                }
+                return length2 - length1;
+            }
+        });
         for (Map.Entry<String, Object> entry : map.entrySet()) {
```
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: nit: replace this loop with
```java
map.forEach(fieldsMap::put);

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 9]
```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
     private void extractFromSource(SearchHit hit)
     {
         Map<String, Object> map = hit.getSourceAsMap();
+        TreeMap<String, Object> fieldsMap = new TreeMap<>(new Comparator<String>()
+        {
+            @Override
+            public int compare(String key1, String key2)
+            {
+                int length1 = key1.split("\\.").length;
```
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: Can we make this a static variable?

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 10]
```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
...
-            Object entryValue = entry.getValue();
-            setFieldIfExists(jsonPath, entryValue);
+            fieldsMap.put(entry.getKey(), entry.getValue());
+        }
+        processFields(fieldsMap);
+    }
+
+    private void processFields(TreeMap<String, Object> fieldsMap)
+    {
+        if (fieldsMap.size() == 0) {
```
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: nit: `fieldsMap.isEmpty()`

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 11]
```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
...
+            return;
+        }
```
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: empty line after this

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 12]
```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
...
+        Map.Entry<String, Object> first = fieldsMap.firstEntry();
+        String field = first.getKey();
+        if (field.contains(".")) {
```
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: also make this static maybe (as a `char` type)

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java (hunk 6 above)
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: nit
```java
fieldsMap.forEach(this::setFieldIfExists);

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 13]
```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
...
+            int size = field.split("\\.").length;
+            Iterator<String> iterator = fieldsMap.navigableKeySet().iterator();
+            while (iterator.hasNext()) {
+                String name = iterator.next();
+                if (name.split("\\.").length == size && name.startsWith(prefix)) {
+                    value.put(name.substring(name.lastIndexOf('.') + 1), fieldsMap.get(name));
+                    iterator.remove();
+                    continue;
+                }
+                break;
```
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: if we break for the first iteration, we don't need a while loop

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java [hunk 14]
```diff
@@ -105,6 +115,6 @@ else if (type.equals(VARCHAR) || type.equals(VARBINARY)) {
         else {
             throw new IllegalArgumentException("Unknown primitive type: " + type.getDisplayName());
         }
-        return builder.build();
+        return;
```
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java]: remove

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java [hunk 15]
```diff
@@ -75,16 +80,21 @@ private static Block serializeStruct(Type type, BlockBuilder builder, Object obj
         }
 
         builder.closeEntry();
-        return (Block) type.getObject(builder, 0);
+        if (builderSynthesized) {
+            return (Block) type.getObject(builder, 0);
+        }
+        else {
```
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java]: else is redundant

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java [hunk 16]
```diff
@@ -46,23 +46,28 @@ public static Block serializeObject(Type type, BlockBuilder builder, Object obje
...
         if (builder == null) {
+            builderSynthesized = true;
             builder = type.createBlockBuilder(null, 1);
         }
-
-        BlockBuilder currentBuilder = builder.beginBlockEntry();
-        List<Type> typeParameters = type.getTypeParameters();
+        currentBuilder = builder.beginBlockEntry();
 
         for (int i = 0; i < typeParameters.size(); i++) {
```
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java]: nit: `for (Type typeParameter : typeParameters) {`

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 17]
```diff
@@ -251,10 +253,53 @@ private Object getFieldValue(int field)
...
+    private void processFields(TreeMap<String, Object> fieldsMap)
+    {
+        if (fieldsMap.size() == 0) {
+            return;
+        }
+        Map.Entry<String, Object> first = fieldsMap.firstEntry();
+        String field = first.getKey();
+        if (field.contains(".")) {
+            String prefix = field.substring(0, field.lastIndexOf('.'));
+            Map<String, Object> value = new HashMap<>();
```
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: use `ImmutableMap.Builder`

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java (hunk 4 above)
Comment (zhenxiao) [reply presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: yes, any levels of nesting is supported
recursion and while loop are processing JSON path with biggest component length, e.g. 
a.b.c and a.b.d will be processed before a.f, and before g

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java (hunk 13 above)
Comment (zhenxiao) [reply presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: need to loop processing all nested json paths with same component length, e.g.
a.b.c.d and a.b.c.f
continue and while are cooperating to make sure json paths with the biggest component length and with the same prefix are processed in each while loop

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java (hunk 5 above)
Comment (zhenxiao) [reply presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: each recursion is to process all nested json paths with same component length, e.g.
step1: a.b.c.d and a.b.c.f
step2: a.b.g and a.b.c
step3: a.b and a.h

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java (hunk 6 above)
Comment (zhenxiao) [reply presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: yes, because TreeMap already sort all json patch based on component length, in descending order

**Comment by zhenxiao** [comment]
kindly ping @wenleix @highker

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 18]
```diff
@@ -251,10 +255,45 @@ private Object getFieldValue(int field)
...
         Map<String, Object> map = hit.getSourceAsMap();
-        for (Map.Entry<String, Object> entry : map.entrySet()) {
-            String jsonPath = entry.getKey();
-            Object entryValue = entry.getValue();
-            setFieldIfExists(jsonPath, entryValue);
+        TreeMap<String, Object> fieldsMap = new TreeMap<>((left, right) -> {
+            int leftLength = left.split(FIELD_DELIMITER_REGEX).length;
+            int rightLength = right.split(FIELD_DELIMITER_REGEX).length;
+            if (leftLength == rightLength) {
+                return left.compareTo(right);
```
Comment (highker) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: The logic is a bit hard to follow because it interleaves string processing with nested field processing. The `compareTo` is a bit weird; for example `a.b` comparing with `ac.d` is going to compare `.` and `c`..
Is it possible to leverage `SubfieldTokenizer` and `Subfield` to separate string processing and nested field processing?

**Comment by dandanmaowu** [comment]
how to config elasticsearch nested object ????

----------------------------------------------------------------------------------------------
elasticsearch object mapping is ok

{
  "id" : "a027b12760cb4568ae911f96621046ea",
  "employee_id" : 26,
  "month_cost" : {
    "month_201010" : "15979.2",
    "month_201011" : "16060"
  }
}


    "columns": [
       {
         "name":"id",
         "type":"varchar",
         "jsonPath":"id",
         "jsonType":"varchar",
         "ordinalPosition":0
       },{
         "name":"employee_id",
         "type":"bigint",
         "jsonPath":"employee_id",
         "jsonType":"bigint",
         "ordinalPosition":1
       },{
         -- this config is right
         "name":"month_cost",
         "type":"row(month_201010 varchar,month_201011 varchar)",
         "jsonPath":"month_cost",
         "jsonType":"varchar",
         "ordinalPosition":2
       }
    ]

----------------------------------------------------------------------------------------------





----------------------------------------------------------------------------------------------
but how to config nested object??? 

    "mappings" : {
      "doc" : {
        "properties" : {
          "name" : {
            "type" : "keyword"
          },
          "month_cost" : {
            "type" : "nested",
            "properties" : {
              "cost" : {
                "type" : "double"
              },
              "month" : {
                "type" : "keyword"
              }
            }
          }
        }
      }
    }



{
  "name" : "zhangsan",
  "month_cost" : [
    {
      "month" : "201901",
      "cost" : 23.22
    },
    {
      "month" : "201902",
      "cost" : 1223.22
    },
    {
      "month" : "201903",
      "cost" : 55.22
    }
  ]
}

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java (hunk 15 above)
Comment (highker) [reply presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java]: else is redundant

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java (hunk 14 above)
Comment (highker) [reply presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java]: remove

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java (hunk 16 above)
Comment (highker) [reply presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java]: nit: `for (Type typeParameter : typeParameters) {`

Comment (zhenxiao) [reply presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchUtils.java]: seems difficult to get fieldName using type. More convenient to get fieldName using TypeSignatureParameter.
